### PR TITLE
Refactor vector calculus operators

### PR DESCRIPTION
**Pull request by danshapero**
This patch pulls all of the specialized vector calculus operators that @benhills added to support flowband models into their own module. I also added functions for the trace of a matrix and to create the Identity matrix. This makes it possible to get rid of the conditionals in the definition of the viscous action for both depth-averaged or 3D models.

### CODE REVIEW

File: icepack/models/hybrid.py [hunk 1]
```diff
@@ -228,9 +214,9 @@ def action(self, **kwargs):
 
         ds_w = ds_v(domain=mesh, subdomain_id=side_wall_ids)
         side_friction = self.side_friction(**kwargs) * ds_w
-        if get_mesh_dimensions(mesh) in ["xy", "xyz"]:
```
Comment (benhills) [inline icepack/models/hybrid.py:231]: I am confused on why we are dropping "xy" here, I could be missing something though.

**Review by benhills — APPROVED** [review]
I really like this. Cleaned up a lot!
I am assuming that you already tested it, but am happy to play around with it if helpful at all.

File: icepack/models/hybrid.py (hunk 1 above)
Comment (danshapero) [reply icepack/models/hybrid.py:231]: The hybrid model only works for "xz" and "xyz". We could add some checks that throw errors early on if you try to use a mesh with the wrong geometry. But if we're going to put effort into sanity-checking inputs there are other places that I think are much more prone to errors.

Comment (benhills) [reply icepack/models/hybrid.py:231]: Duh... ha, thanks.

Comment (danshapero) [reply icepack/models/hybrid.py:231]: No worries! If it makes you feel any better, the reviewers for the icepack paper *hated* me calling this a "hybrid" model, so apparently it's confusing to everyone.

File: icepack/models/friction.py [hunk 2]
```diff
@@ -77,8 +76,7 @@ def normal_flow_penalty(**kwargs):
     scale = kwargs.get("scale", firedrake.Constant(1.0))
 
     mesh = u.ufl_domain()
```
[the remark this answers is not in the document]
Comment (danshapero) [reply icepack/models/friction.py:78]: My usual approach is to keep the assignment if the variable gets used in several places and get rid of it if it's used in only one. In the previous function to calculate the side wall friction, the mesh was only used in one place, so I took out the assignment to a variable for it. In the normal flow penalty function, you have to pass the mesh around to get the facet normal vectors and facet areas as well as the diameter of the domain.

[the remark this answers is not in the document]
Comment (benhills) [reply icepack/models/friction.py:78]: All good! This has been more of a learning experience for me than it has been a 'review' but that is what I had to offer I guess ;).
Thanks for looping me in Daniel!

**Pull request merged**